config: also use XDG_CONFIG_HOME on MacOS

ckoehler opened the pull request:
This replaces the `dirs` crate with the more lightweight and flexible `etcetera` crate, which also lets us use XDG config on MacOS.


Fixes #3434.

~~I'm not entirely sure how do prevent the line break in the doc table between "Application" and "Support" 😅 ...~~

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

noahmayr commented on docs/config.md:
```diff
@@ -745,10 +745,10 @@ platform-specific location is recommended for better integration with platform
-| Windows  | `{FOLDERID_RoamingAppData}\jj\config.toml`         | `C:\Users\Alice\AppData\Roaming\jj\config.toml`           |
+| :------- | -------------------------------------------------- | --------------------------------------------------------- |
+| Linux    | `$XDG_CONFIG_HOME/jj/config.toml`                    | `/home/alice/.config/jj/config.toml`                        |
+| macOS    | `$HOME/Library/Application Support/jj/config.toml` or `XDG_CONFIG_HOME/jj/config.toml` as fallback | `/Users/Alice/Library/Application Support/jj/config.toml` or `/Users/Alice/.config/jj/config.toml` |
```
Nit: you're missing the dollar sign of `$XDG_CONFIG_HOME`

noahmayr commented on cli/src/config.rs:
```diff
@@ -434,6 +435,41 @@ fn read_config_path(config_path: &Path) -> Result<config::Config, config::Config
+        }
+    };
+
+    // Try getting config directory from Apple base strategy. `Application Support`
+    // is really the data_dir, so get that.
+    let path = etcetera::base_strategy::Apple::new()
+        .ok()
+        .and_then(|strategy| check_config(&strategy.data_dir()));
+
+    // If config directory is not found from Apple base strategy, try XDG instead
+    path.or_else(|| {
+        etcetera::choose_base_strategy()
+            .ok()
+            .and_then(|strategy| check_config(&strategy.config_dir()))
+    })
```
I think we should prefer XDG's strategy over Apple so new users get this behavior by default when running `jj config set` or `jj config edit` without having an existing config in `~/Library/Application Supprt/`

noahmayr commented:
> I'm not entirely sure how do prevent the line break in the doc table between "Application" and "Support" 😅 ...

I don't think there's a line break, it's just Github's wrapping (check the line numbers on the left, the second line doesn't have one)

noahmayr commented on cli/src/config.rs:
```diff
@@ -434,6 +435,41 @@ fn read_config_path(config_path: &Path) -> Result<config::Config, config::Config
+
+    // If config directory is not found from Xdg, try Apple instead. `Application
+    // Support` is really the data_dir, so get that.
+    path.or_else(|| {
+        etcetera::base_strategy::Apple::new()
+            .ok()
+            .and_then(|strategy| check_config(&strategy.data_dir()))
+    })
```
Hmm, actually this will result in the config dir still being `~/Library/Application Support/` if you don't have a config yet.

ckoehler replied:
Hm I can't reproduce that. On my machine, if I don't have a config either place, it creates `$HOME/.jjconfig.toml`, which is probably not what we want either.

ilyagr commented on cli/src/config.rs:
```diff
@@ -434,6 +435,41 @@ fn read_config_path(config_path: &Path) -> Result<config::Config, config::Config
+    };
+
+    // Try getting config directory from Xdg base strategy first.
+    let path = etcetera::base_strategy::Xdg::new()
```
We probably should report an error if both the Apple config file and the XDG config file exist. This is what we do if both `~/.jjconfig` and the Apple config file exist. It'll require people to remove one of them, but that's probably better than them thinking that they are using one file while they are actually using another one.

ilyagr replied:
Alternatively (and I'm not sure this is what we should do), we could *only* report an error if the Apple path exists, and always use the XDG path of `.jjconfig.toml`, like on Linux. I'm not sure what should happen on Windows, but that's outside the scope of this PR anyway.

ckoehler replied:
I like the idea of only having one location on MacOS in addition to `$HOME/.jjconfig.toml`. Error on finding one in Application Support, and only use the XDG one. Is that acceptable?

ilyagr replied:
I think that would work, unless somebody feels there's some advantage to following the Apple way.

You'd have to mark it as a breaking change in the changelog. We'd also want to keep a lookout for confused or unhappy users after merging that.

joyously replied:
Is it bad form to move the file if it exists in the old location? Like cleaning up from old versions...

ckoehler replied:
> Is it bad form to move the file if it exists in the old location? Like cleaning up from old versions...

Hm I'd find that confusing and would rather do it myself.

> You'd have to mark it as a breaking change in the changelog. We'd also want to keep a lookout for confused or unhappy users after merging that.

I'll mark it breaking.

ilyagr replied:
It might be better to have the user move it, just to make sure they know what changed. We could print the `mkdir -p && mv` command for them, though.

ckoehler replied:
I went with the last one.

chriskrycho commented on CHANGELOG.md:
```diff
@@ -11,6 +11,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Breaking changes
 
+* On MacOS, jj will now look for its config in `$XDG_CONFIG_HOME` and error if config is found in `~/Library/Application Support/jj/`
```
I thought, when I originally saw this, that the idea was to *add* support for `$XDG_CONFIG_HOME` as a configurable default, but not *remove* support for the OS default. I am just one guy, but I *really* prefer things to go in the OS’ preferred locations instead of in `~/.config`. I have kind of resigned myself to having a bunch of junk in `~/.config`, but I am a Mac user first and foremost, and do not share a lot of the XDG philosophy or any commitment to it.

My own *very* strong preference here (and people who actually do work maintaining jj should of course just overrule this, but I wanted to flag it up) is to support both on macOS, given that I know lots of people *do* prefer the XDG location over `~/Library/Application Support`. There is a design question there about what the default should be, given a lot of users *do* use macOS as a nice UI and good hardware but fundamentally as a \*nix.

To make the point another way: imagine doing this to Windows users and telling them that `%APPDATA` was deprecated and they needed to just deal with having the XDG config added to `C:\Documents and Settings\chris\.config`. That… would go over badly. 😂 As it should!

ckoehler replied:
Fair, I'll keep Application Support as a fallback.

ckoehler commented:
Sooo I'm finding the current behavior of the config file handling really convoluted. As a result, I am making bigger changes than I thought I'd have to. Before I go too far (maybe it's too late already!), I'd like some feedback:
- Is this okay? Too much? 
- Do the tests describe the wanted behaviors well?

The commits themselves are kind of a mess (I want to keep more history with so many changes), so don't look at those yet. I'll clean them up before I merge anything.

Thanks!

ilyagr commented:
> Before I go too far (maybe it's too late already!), I'd like some feedback:
> 
>    - Is this okay? Too much?
>    - Do the tests describe the wanted behaviors well?
> 
> The commits themselves are kind of a mess 

I'm a little confused -- what are the changes (especially if the commits are too raw to be looked at)?

Didn't we decide that we should allow all three locations?

hypatian commented:
I'm very confused here—from reading through the changes, it looks to me like despite the documentation and the commit message claiming that on macOS it will use XDG_CONFIG_DIR by default, it will actually use the system location by default. And, of course, it gives errors on multiple locations not on any specific location.

Has the documentation (including commit message) not been updated to match the code changes?

ckoehler commented:
> Didn't we decide that we should allow all three locations? At least the commit descriptions seem to suggest the PR still forbids the `Application Support` location.

Yes, all three are allowed. I’ll clean it up later. Sounds like the most helpful thing would be to describe the behavior this introduces.

> I'm very confused here—from reading through the changes, it looks to me like despite the documentation and the commit message claiming that on macOS it will use XDG_CONFIG_DIR by default, it will actually use the system location by default. And, of course, it gives errors on multiple locations not on any specific location. (Edit: to be clear: it will _read_ from the system location by default. Newly created files go to ~/.jjconfig.toml, it looks like.)

Yes, if there is no configuration yet and jj has to create it, it goes into $HOME. If more than one exists, it shows an error. If only one exists, it uses that. For Mac, that can be Application Support, $XDG_CONFIG_HOME, or $HOME/.jjconfig.toml.

> Has the documentation (including commit message) not been updated to match the code changes?

Not yet. I’ll do that.

gulbanana commented:
I don't think it's desirable to put the new config in $HOME instead of the OS-specific location by default, basically for the reasons here https://github.com/martinvonz/jj/pull/3466#discussion_r1557987601. The problem with the existing config implementation was that it didn't _support_ XDG_CONFIG_HOME; adding that support is good, but we shouldn't break the use of AppData/Application Support for people who aren't using XDG. Which will be most people.

ckoehler commented:
I updated all the docs and changelog and cleaned up the commit. The behavior of some of the config commands creating config files if none exist is really separate from this, so it can be its own PR.

martinvonz reviewed (changes requested):
I find this hard to review. It feels like it does many things at once. Is it possible to split it up something like the following?

* change from `dirs` to `etcetera`
* refactor tests
* refactor `config_path()`
* add support for XDG_CONFIG_HOME

ckoehler commented:
> I find this hard to review. It feels like it does many things at once. Is it possible to split it up something like the following?
> 
>     * change from `dirs` to `etcetera`
> 
>     * refactor tests
> 
>     * refactor `config_path()`
> 
>     * add support for XDG_CONFIG_HOME

Sure! Those individual commits just won't work by themselves. I interpreted atomic as "need to build and work independently".

martinvonz commented:
> I interpreted atomic as "need to build and work independently".

Yes, that's correct.

> Those individual commits just won't work by themselves.

Why not?

ckoehler commented:
> > Those individual commits just won't work by themselves.
> 
> Why not?

Eh maybe I spoke too soon. Let me give it a shot. Thanks for the feedback :)

ckoehler commented:
Alright, I hope this is easier to review.

martinvonz commented on cli/src/config.rs:
> config: refactor config_path() to be clearer

This could have used more detail. How is refactored? If I've deciphered the code changes correctly, it does this:

* renamed `config_path()` to `actual_configs()`. Maybe `existing_config_path()` would be a better name?
* Take `self` by reference
* Change return type from `ConfigPath` to `Option<PathBuf>`, maybe by merging `New` and `Unavailable` to `None`?
* Changed it so we no longer create a file in `~/.jjconfig.toml` when there's no platform config. Was this intended?
* Something changed in `existing_config_path()` but I'm not sure what or why
* There's some new code at the start of `new_or_existing_config_path()`. I'm not sure what that's about either. Is it a required because of the other refactorings?

It might make sense to split up the commit more in some way. Then you'll get a better chance to explain the motivation for each commit when it's non-obvious.

ckoehler replied:
I simplified the refactor. 

> * Changed it so we no longer create a file in `~/.jjconfig.toml` when there's no platform config. Was this intended?

Yes, etcetera returns an error if it can't give a path for the platform dir, so I rely on that to tell the user. That happens pretty early in the process.

martinvonz reviewed (commented):
Thanks, it was a lot easier to review now, but I'm still quite confused about the third commit.

martinvonz commented on cli/src/config.rs:
```diff
@@ -266,62 +254,84 @@ impl ConfigEnv {
+        if let Some(config) = existing_configs? {
+            Ok(Some(config))
+        // otherwise, create one and then return it
+        } else if matches!(handle_missing, HandleMissingConfig::Create) {
```
nit: i think `handle_missing == HandleMissingConfig::Create` is more idiomatic

martinvonz replied:
Oh, but you can probably just add `#[derive(PartialEq)]` on the type, right?

ckoehler commented on cli/src/config.rs:
```diff
@@ -249,12 +252,18 @@ struct ConfigEnv {
+            home_dir: etcetera::home_dir().ok(),
             jj_config: env::var("JJ_CONFIG").ok(),
-        }
+        })
```
The problem is that if one fails, both will, because they both hinge on finding the home directory. That's true for the existing code, so IIUC there was never a situation where it picked $HOME over platform dir, and remains true for this one. I'd prefer to know about that (unlikely as it is, since it doesn't just look for $HOME but also uses some system calls to figure it out).
But if you prefer I can just return None.

martinvonz replied:
> The problem is that if one fails, both will, because they both hinge on finding the home directory.

Is that true if `$XDG_CONFIG_HOME` points outside of `$HOME`? Anyway, I think you're right that it's better to error out if we can't find the config directory. Assuming that's what we decide, shouldn't that be true whether we use `dirs` or `etcetera`? I.e., I don't think we should combine the change in behavior with the change from `dirs` to `etcetera`. Does that make sense?

ckoehler replied:
> Is that true if `$XDG_CONFIG_HOME` points outside of `$HOME`? 

You're right, in that instance it isn't true.

> Anyway, I think you're right that it's better to error out if we can't find the config directory. Assuming that's what we decide, shouldn't that be true whether we use `dirs` or `etcetera`? I.e., I don't think we should combine the change in behavior with the change from `dirs` to `etcetera`. Does that make sense?

I think you're saying that we should have etcetera not error, but convert to `None` instead, and continue having `ConfigEnv::new` return just `Self` instead of a `Result`? Then if we want to have it error instead, we can do that later?

martinvonz replied:
> I think you're saying that we should have etcetera not error, but convert to `None` instead, and continue having `ConfigEnv::new` return just `Self` instead of a `Result`? Then if we want to have it error instead, we can do that later?

Yes, that seems like the safest bet (least controversial). The other option is to instead start by changing the behavior while still using `dirs` in a first commit (or several if it gets complicated), and then switch to `etcetera` in a later commit. The risk with that approach is that people will object and it ends up blocking the later changes you want to make (adding support for `XDG_CONFIG_HOME` on MacOS).

ckoehler replied:
Yeah, and `dirs` doesn't return errors anyway, so not sure how to handle that (turn None into an Err??). I'll make the change.

ckoehler replied:
Alright, done.

ckoehler commented on cli/src/config.rs:
I mentioned that above, but the only change in behavior is that it reports an error if it can't find the home directory.

martinvonz commented on cli/src/config.rs:
```diff
@@ -268,62 +257,84 @@ impl ConfigEnv {
             // TODO: We should probably support colon-separated (std::env::split_paths)
-            return Ok(ConfigPath::new(Some(PathBuf::from(path))));
+            let path = PathBuf::from(path);
+            if handle_missing == HandleMissingConfig::Create && !path.exists() {
+                create_config_file(&path)?;
+            }
```
Was this required because of some other changes? Or does this represent a change in behavior?

ckoehler replied:
I combined the two functions into one, with an argument to create the file when needed. This is existing behavior.

yuja replied:
Perhaps, `create_config_file()` can be moved to caller? It's generally easier to test if the function had no side effect.

martinvonz commented on cli/src/config.rs:
```diff
@@ -268,62 +257,84 @@ impl ConfigEnv {
-        let platform_config_path = ConfigPath::new(self.config_dir.map(|mut config_dir| {
+        // look for existing configs first
+        let platform_config = self.config_dir.as_ref().map(|config_dir| {
+            let mut config_dir = config_dir.clone();
```
Can this look more similar to the old code like this?

```rust
self.config_dir.map(|mut config_dir| {
```

martinvonz commented on cli/src/config.rs:
```diff
@@ -268,62 +257,84 @@ impl ConfigEnv {
-            (Existing(platform_config_path), Existing(home_config_path)) => Err(
-                ConfigError::AmbiguousSource(platform_config_path, home_config_path),
-            ),
-            (Existing(path), _) | (_, Existing(path)) => Ok(Existing(path)),
```
In the commit message, you write this:

> * if told to create missing, it creates one at the platform config location (not $HOME)

It looks to me like this line made it so the old behavior already was that. Am I missing something or is the commit message incorrect (maybe stale)?

ckoehler replied:
You're right, I'll update the description.

yuja commented on cli/src/config.rs:
```diff
@@ -250,9 +251,19 @@ struct ConfigEnv {
+        #[cfg(target_os = "macos")]
+        let platform_config_dir = etcetera::base_strategy::Apple::new()
+            .ok()
+            .map(|c| c.data_dir());
```
nit: `if cfg!(..) { .. } else { .. }` can be used

yuja commented on cli/src/config.rs:
```diff
@@ -958,25 +960,31 @@ mod tests {
-                Want::New(want) => (None, Some(want)),
-                Want::ExistingAndNew(want) => (Some(want), Some(want)),
+            match self.want {
+                Want::None => {}
```
Doesn't it remove the assertion for `Want::None` case?

I don't remember the original expectation of these tests, but it doesn't make sense to run a test with no assertion.

ckoehler replied:
Thanks for pointing that out, you're exactly right! The `Want::None` test doesn't really make sense, because we need to either call the function that gives us a config path without creating it or with creating it. 

What we want is a way to tell the test to run one or the other, and what to expect out of each. I'm changing it accordingly.

martinvonz replied:
It looks like this was added in #1861, so maybe @qfel can clarify?

qfel replied:
It has been a while, so I don't remember well, but..

It looks like Want::None used to declare both new_config_path and existing_config_path were expected to return None. In this PR test_config_path_home_dir_doesnt_create_new uses Want::None. What do you want to assert for that test? That both existing_config_path and new_or_existing_config_path return None, or something else?

ckoehler replied:
I didn't understand the behavior too well before. In this case it should be None with `existing_config_path` and `Some("<path>")` with `new_or_existing_config_path`.

yuja commented on cli/src/config.rs:
```diff
@@ -31,6 +31,8 @@ pub enum ConfigError {
     #[error("Both {0} and {1} exist. Please consolidate your configs in one of them.")]
     AmbiguousSource(PathBuf, PathBuf),
+    #[error("Configs found in {0}, {1}, and {2}. Please consolidate your configs in one of them.")]
+    AmbiguousSource3(PathBuf, PathBuf, PathBuf),
```
Perhaps, these two error variants can be combined as `AmbiguousSource(Vec<PathBuf>)`. The error message could be formatted as something like `"Configs found in multiple locations: {}", .0.iter().map(|path| path.display()).join(", ")` (untested)

ckoehler commented:
I'm going to take a break from this PR, it's been taking a lot more time than I expected. Someone else is free to take it over.

PhilipMetzger commented:
I'm closing this as #6300 now has landed in main.
